File: src/main/java/com/graqr/threshr/model/queryparam/Place.java
package com.graqr.threshr.model.queryparam;

import io.micronaut.serde.annotation.Serdeable;
import lombok.Data;

/**
 * Place is a query parameter used when querying nearby target store locations. The query accepts either a zipcode
 * or a city-state combo. See {@link com.graqr.threshr.Threshr#queryStoreLocations(Place)}.
 */
@Serdeable
@Data
public class Place {
    private final String zipOrCityState;
    public Place(String zipcode) {
        if (!String.valueOf(zipcode).matches("^\\d{5}[-|\\s]?(\\d{4})?$")) {
            throw new IllegalArgumentException("Invalid zipcode provided, \"" + zipcode + "\". Zipcode provided " +
                    "must match this regex: \"^\\d{5}(-|\\s*)?(\\d{4})?$\"");
        }
        zipOrCityState = zipcode;
    }

    public Place(String city, String state) {
        String errorMessage = "";
        //match weird strings like D'Amoreport (texas) and Coeur D'Alene (idaho)
        String pattern = "^([a-zA-Z|\\u0080-\\u024F]\\W?)*+$"; // *+ quantifier prevents backtracking
        if (!String.valueOf(city).matches(pattern)) {
            errorMessage = "Invalid city string provided, \"" + city + "\".";
        } else if (city.isEmpty()) {
            errorMessage = "String value for city cannot be empty.";
        }
        if (!String.valueOf(state).matches(pattern)) {
            errorMessage = "Invalid state string provided, \"" + state + "\".";
        } else if (state.isEmpty()) {
            errorMessage += "String value for state cannot be empty.";
        }
        if (!errorMessage.isEmpty()) {
            throw new IllegalArgumentException(errorMessage);
        }
        zipOrCityState = String.format("%s, %s", city, state);
    }
}

File: src/main/kotlin/com/graqr/threshr/circleStore/model/queryparam/Place.kt
package com.graqr.threshr.circleStore.model.queryparam

/**
 * This class represents a location in the USA. It is used by the RedSky API's Nearby Stores endpoint, which
 * accepts a location in the form of a zipcode or a city, state combo. The Nearby Stores endpoint will return a
 * list of stores near the given location.
 *
 * @param zipOrCity either a zipcode or a city and state provided as two separate strings
 * @author Jonathan Zollinger
 * @since 0.1.0
 */
class Place(val zipOrCity: String) {
    /*
     * Initializes a Place object using a zipcode.
     *
     * This constructor is used when the location is provided as a zipcode.
     * The zipcode must be a 5-digit number and can optionally include a 4-digit extension,
     * separated by a hyphen or space.
     */
    init {
        require(zipOrCity matches Regex("^\\d{5}[-|\\s]?(\\d{4})?$")) {
            "Invalid zipcode provided. Zipcode must be a 5-digit number optionally followed by a 4-digit extension."
        }
    }


    /**
     * Initializes a Place object using a city and state.
     *
     * @param city city name
     * @param state state name
     */
    constructor(city: String, state: String) : this("$city, $state") {
        //match weird strings like D'Amoreport (texas) and Coeur D'Alene (idaho)
        val pattern = Regex("^([a-zA-Z|\\u0080-\\u024F]\\W?)*+$")
        require(city matches pattern) {
            "Invalid city string provided, \"$city\". City must contain only letters and whitespace."
        }
        require(state matches pattern) {
            "Invalid state string provided, \"$state\". State must contain only letters and whitespace."
        }
    }
}
